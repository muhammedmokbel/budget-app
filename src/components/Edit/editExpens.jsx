import React from 'react'

import ExpenseForm from '../ExpenseForm/ExpenseForm'

import {connect} from 'react-redux'

import {editExpense , removeExpense} from '../../redux/actionGenerators/expenseGenerators'


export class EditExpense extends React.Component {

    render() {
        return (
            <div>
                edit page
            <ExpenseForm 
            onSubmit={(expenseData) => {
                this.props.editExpense(this.props.match.params.id , expenseData)
                this.props.history.push('/')
            }} 
            {...this.props.expenses.filter(item => item.id === this.props.match.params.id )[0]}/>
        
            <button onClick={() =>{ 
                this.props.removeExpense( this.props.match.params.id)
                this.props.history.push('/')
            }}>Remove</button>
            </div>
        )
    }
} 


const mapStateToProps = state => {
    return {
        expenses : state.expenseReducer.expenses
    }
}

const mapDisptachToProps = (dispatch, props) => {
    return {
        editExpense : (id , val) => dispatch(editExpense(id , val)) , 
        removeExpense : val => dispatch(removeExpense({id : val}))
    }
}

export default connect(mapStateToProps, mapDisptachToProps)(EditExpense);