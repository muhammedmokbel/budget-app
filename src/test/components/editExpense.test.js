import moment from 'moment'
import React from 'react' 
import { shallow } from 'enzyme'
import {EditExpense} from '../../components/Edit/editExpens'


const expensesList =  [{
    id: '1',
    description: 'Gum',
    note: '',
    amount: 195,
    createdAt: 0
  }, {
    id: '2',
    description: 'Rent',
    note: '',
    amount: 109500,
    createdAt: moment(0).subtract(4, 'days').valueOf()
  }, {
    id: '3',
    description: 'Credit Card',
    note: '',
    amount: 4500,
    createdAt: moment(0).add(4, 'days').valueOf()
  }];

let wrapper , match , history , editExpense , removeExpense; 

beforeEach(() => {
    editExpense = jest.fn()
    removeExpense = jest.fn()
    match = {
        params : {
            id : '3'
        }
    }
    history = { push : jest.fn()}
    wrapper = shallow(<EditExpense removeExpense={removeExpense} editExpense={editExpense} expenses={expensesList}  match={match} history={history} />)
})

test('should render edit expense component correctly' , () => {
    expect(wrapper).toMatchSnapshot()
})

test('should return edit data correctly' , () => {

    wrapper.find('ExpenseForm').prop('onSubmit')(expensesList[2])

    expect(editExpense).toHaveBeenCalledWith('3' , expensesList[2])
    expect(history.push).toHaveBeenCalledWith('/')
    
})

test('should return id of removed data correctly', () => {

    wrapper.find('button').simulate('click')

    expect(removeExpense).toHaveBeenCalledWith('3')
    expect(history.push).toHaveBeenCalledWith('/')

})