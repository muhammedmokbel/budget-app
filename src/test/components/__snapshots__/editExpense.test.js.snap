// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense component correctly 1`] = `
<div>
  edit page
  <ExpenseForm
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    note=""
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
